core/src/main/scala/io/delta/tables/DeltaOptimizeBuilder.scala

```diff
@@ -47,7 +47,12 @@ class DeltaOptimizeBuilder private(
    * @since 2.0.0
    */
   def where(partitionFilter: String): DeltaOptimizeBuilder = {
-    this.partitionFilter = Some(partitionFilter)
+    if (this.partitionFilter.isEmpty) {
+      this.partitionFilter = Some(partitionFilter)
+    }
+    else {
+      this.partitionFilter = Some(this.partitionFilter.getOrElse() + " and " + partitionFilter)
```

Comment:
ohhh.. this sort of streaming SQL string manipulations is generally unsafe and tends to have horrible complications found years and months later. For example, have thought about operator precedence?

In general, there are 2 valid ways of dealing wit multiple where calls. 
1. we do the concatenation using parsed logical expressions objects. This is how Spark does it. When there are multiple `dataframe.where("a=1").where("x=2")`  it immediately boils down to the 
```Filter(Equals(x, 2), 
        Filter(Equals(a, 1), 
           <logical-plan-of-dataframe>))
``` which eventually gets optimized to `Filter(And(Equals(x, 2), Equals(a, 1)), ...)`

2. we dont support multiple where, and we throw explicit error `where()` is called twice to avoid confusing behavior of things getting silently ignored.

I think 1 may be hard to do without having access to the SQL parser in this location.... so I think 2 is better for now.

Reply:
thinking a bit more about operator precedence... what if we put each filter inside `(...)`.... so `(filter1) AND (filter2)`?
@zsxwing what do you think ... will that make it safe?

Reply:
I think we can store these strings as a seq, and then parse them in `io.delta.tables.DeltaOptimizeBuilder.execute` to get a seq of filters and AND them.

Reply:
ohhh thats good. i just noticed. thank you. okay.. then thats the way to go.

Reply:
@tdas  just to confirm something like this ?

where function will just add to seq 

and in execute we change it to something like this . 
```
  private def execute(zOrderBy: Seq[UnresolvedAttribute]): DataFrame = {
    val tableId: TableIdentifier = sparkSession
      .sessionState
      .sqlParser
      .parseTableIdentifier(tableIdentifier)
    val optimize = OptimizeTableCommand(None, Some(tableId), partitionFilter.mkString(" AND "))(zOrderBy = zOrderBy)
    toDataset(sparkSession, optimize)
  }
}
```
or i am missing something ?

Reply:
You can parse the strings down to https://github.com/delta-io/delta/blob/d23baecc2eb93c23d44b4ea79d9778f6a67c77ac/core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala#L125-L134 and get Expression objects for each partition filter.